pkg/controller/labelidentity/label_group_index.go

```diff
@@ -0,0 +1,504 @@
+	nsLabels := constructMapFromLabelString(labelMatches[nsIndex])
+	podLabels := constructMapFromLabelString(labelMatches[podIndex])
+
+	namespace := nsLabels[apiv1.LabelMetadataName]
+	return &labelIdentityMatch{
+		id:               id,
+		namespace:        namespace,
+		namespaceLabels:  nsLabels,
+		podLabels:        podLabels,
+		selectorItemKeys: sets.NewString(),
+	}
+}
+
+// selectorItemKeyFunc knows how to get the key of a selectorItem.
+func selectorItemKeyFunc(obj interface{}) (string, error) {
+	sItem, ok := obj.(*selectorItem)
+	if !ok {
+		return "", fmt.Errorf("object is not of type *selectorItem: %v", obj)
+	}
+	return sItem.getKey(), nil
+}
+
+func newSelectorItemStore() cache.Indexer {
+	indexers := cache.Indexers{
+		cache.NamespaceIndex: func(obj interface{}) ([]string, error) {
+			sItem, ok := obj.(*selectorItem)
+			if !ok {
+				return []string{}, nil
+			}
+			// sItem.Selector.Namespace == "" means it's a cluster scoped selector, we index it as it is.
+			return []string{sItem.selector.Namespace}, nil
+		},
+		policyIndex: func(obj interface{}) ([]string, error) {
+			sItem, ok := obj.(*selectorItem)
+			if !ok {
+				return []string{}, nil
+			}
+			return sItem.policyKeys.UnsortedList(), nil
+		},
+	}
+	return cache.NewIndexer(selectorItemKeyFunc, indexers)
+}
+
+// LabelIdentityIndex implements Interface.
+type LabelIdentityIndex struct {
+	lock sync.RWMutex
+	// labelIdentities stores all labelIdentityMatches, with the normalized labels of LabelIdentity as map key.
+	labelIdentities map[string]*labelIdentityMatch
+	// labelIdentityNamespaceIndex is an index from Namespace to LabelIdentity keys in that Namespace.
+	labelIdentityNamespaceIndex map[string]sets.String
+	// selectorItems stores all selectorItems, indexed by Namespace and policy keys.
+	selectorItems cache.Indexer
+
+	eventChan chan string
+	// eventHandlers is a list of callbacks registered for policies to be re-processed due to
+	// LabelIdentity events.
+	eventHandlers []eventHandler
+
+	// synced stores a boolean value, which tracks if the LabelIdentityIndex has been initialized with
+	// the full lists of LabelIdentities.
+	synced *atomic.Value
+}
+
+func NewLabelIdentityIndex() *LabelIdentityIndex {
+	synced := &atomic.Value{}
+	synced.Store(false)
+	index := &LabelIdentityIndex{
+		labelIdentities:             map[string]*labelIdentityMatch{},
+		labelIdentityNamespaceIndex: map[string]sets.String{},
+		selectorItems:               newSelectorItemStore(),
+		eventChan:                   make(chan string, eventChanSize),
+		eventHandlers:               []eventHandler{},
+		synced:                      synced,
+	}
+	return index
+}
+
+func (i *LabelIdentityIndex) updateSelectorItem(sItem *selectorItem, updateType selectorItemUpdateEvent, updateKey string) {
+	// Make a copy of selectorItem's policy keys as modifying the original could affect indexing.
+	labelIdentities, policies := sItem.labelIdentityKeys, sets.StringKeySet(sItem.policyKeys)
```

Comment:
labelIdentities needs to be a copy as it's also updated, reason is as below comment.

A quick way to make a copy of a set is `sItem.policyKeys.Union(nil)`

Reply:
Fixed